{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script produces a table of PMC paper IDs matched to the SRA or GEO datasets they contain, and metadata about those datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required modules\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rc\n",
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function returning corresponding metadata or Study/Series ID given any information on a dataset\n",
    "def grabRelated(table, key, key_col, out_col):\n",
    "    indices = table.index[table[key_col] == key].tolist()\n",
    "    if(len(indices) > 0):\n",
    "        allHits = table.loc[indices][[out_col]]\n",
    "        modeHits = allHits.mode()\n",
    "        if(modeHits.empty):\n",
    "            return 'NaN'\n",
    "        else:\n",
    "            return modeHits.iloc[0,0]\n",
    "    else:\n",
    "        return 'NaN'\n",
    "\n",
    "# define functions to perform QC on accessions, checking if they exist in the references\n",
    "def sraChecksOut(acc, col):\n",
    "    if(acc in sraReference[col].tolist()):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "def geoChecksOut(acc, col):\n",
    "    if(acc in geoReference[col].tolist()):\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "STEP I: Clean the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Series</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Datasets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GSM2</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE462</td>\n",
       "      <td>GSM3</td>\n",
       "      <td>GPL5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GSE462</td>\n",
       "      <td>GSM4</td>\n",
       "      <td>GPL5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GSE462</td>\n",
       "      <td>GSM5</td>\n",
       "      <td>GPL5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3280011</td>\n",
       "      <td>GSE136775</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3280012</td>\n",
       "      <td>GSE136776</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3280013</td>\n",
       "      <td>GSE137458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3280014</td>\n",
       "      <td>GSE137562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3280015</td>\n",
       "      <td>GSE137563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3280016 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Series Accession Platform Datasets\n",
       "0           GSE506      GSM1     GPL4      NaN\n",
       "1           GSE506      GSM2     GPL4      NaN\n",
       "2           GSE462      GSM3     GPL5      NaN\n",
       "3           GSE462      GSM4     GPL5      NaN\n",
       "4           GSE462      GSM5     GPL5      NaN\n",
       "...            ...       ...      ...      ...\n",
       "3280011  GSE136775       NaN      NaN      NaN\n",
       "3280012  GSE136776       NaN      NaN      NaN\n",
       "3280013  GSE137458       NaN      NaN      NaN\n",
       "3280014  GSE137562       NaN      NaN      NaN\n",
       "3280015  GSE137563       NaN      NaN      NaN\n",
       "\n",
       "[3280016 rows x 4 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import GEO reference data\n",
    "\n",
    "# import table with accession, platform, and series\n",
    "geoAPS = pd.read_csv('../geo_sample.csv')\n",
    "\n",
    "# import table with datasets IDs\n",
    "geoDS = pd.read_csv('../geo_series_table.csv', low_memory = False)\n",
    "geoDS.columns = ['Series', 'Title', 'Series Type', 'Taxonomy', \n",
    "                 'Sample Count', 'Datasets', 'Supplementary Types', \n",
    "                 'Supplementary Links', 'PubMed ID', 'SRA Accession', \n",
    "                 'Contact', 'Release Date']\n",
    "\n",
    "# add datasets column by merging\n",
    "allData = pd.merge(geoAPS, geoDS, how = 'outer', on = 'Series')\n",
    "geoReference = allData[['Series', 'Accession', 'Platform', 'Datasets']]\n",
    "\n",
    "geoReference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRAStudy</th>\n",
       "      <th>Run</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Submission</th>\n",
       "      <th>Sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>ERR3550121</td>\n",
       "      <td>ERX3556726</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>ERS3773255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>ERR3550142</td>\n",
       "      <td>ERX3556747</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>ERS3773255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>ERR3550147</td>\n",
       "      <td>ERX3556752</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>ERS3773263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>ERR3550125</td>\n",
       "      <td>ERX3556730</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>ERS3773263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>ERR3550137</td>\n",
       "      <td>ERX3556742</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>ERS3773252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676058</td>\n",
       "      <td>SRP002493</td>\n",
       "      <td>SRR049674</td>\n",
       "      <td>SRX020916</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRA017695</td>\n",
       "      <td>SRS073299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676059</td>\n",
       "      <td>SRP002493</td>\n",
       "      <td>SRR049675</td>\n",
       "      <td>SRX021028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRA017695</td>\n",
       "      <td>SRS073299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676060</td>\n",
       "      <td>SRP002493</td>\n",
       "      <td>SRR049677</td>\n",
       "      <td>SRX021029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRA017695</td>\n",
       "      <td>SRS073299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676061</td>\n",
       "      <td>SRP002064</td>\n",
       "      <td>SRR042496</td>\n",
       "      <td>SRX020084</td>\n",
       "      <td>PRJNA46359</td>\n",
       "      <td>DIEGO</td>\n",
       "      <td>SRS011854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676062</td>\n",
       "      <td>SRP002438</td>\n",
       "      <td>SRR043092</td>\n",
       "      <td>SRX020242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OF</td>\n",
       "      <td>SRS072567</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7676063 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          SRAStudy         Run  Experiment  BioProject  Submission      Sample\n",
       "0        ERP000767  ERR3550121  ERX3556726   PRJEB2600  ERA2143351  ERS3773255\n",
       "1        ERP000767  ERR3550142  ERX3556747   PRJEB2600  ERA2143351  ERS3773255\n",
       "2        ERP000767  ERR3550147  ERX3556752   PRJEB2600  ERA2143351  ERS3773263\n",
       "3        ERP000767  ERR3550125  ERX3556730   PRJEB2600  ERA2143351  ERS3773263\n",
       "4        ERP000767  ERR3550137  ERX3556742   PRJEB2600  ERA2143351  ERS3773252\n",
       "...            ...         ...         ...         ...         ...         ...\n",
       "7676058  SRP002493   SRR049674   SRX020916         NaN   SRA017695   SRS073299\n",
       "7676059  SRP002493   SRR049675   SRX021028         NaN   SRA017695   SRS073299\n",
       "7676060  SRP002493   SRR049677   SRX021029         NaN   SRA017695   SRS073299\n",
       "7676061  SRP002064   SRR042496   SRX020084  PRJNA46359       DIEGO   SRS011854\n",
       "7676062  SRP002438   SRR043092   SRX020242         NaN          OF   SRS072567\n",
       "\n",
       "[7676063 rows x 6 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import SRA reference data\n",
    "sraReference = pd.read_csv('../sraIDfull.csv', error_bad_lines = False, low_memory=False, quoting=3)\n",
    "sraReference = sraReference[['SRAStudy', 'Run', 'Experiment', \n",
    "                             'BioProject', 'Submission', 'Sample']]\n",
    "sraReference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144401</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144402</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144403</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144404</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144405</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144406 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    journal      pmc_ID    accession\n",
       "0       Alzheimers_Res_Ther  PMC3707052         GSM1\n",
       "1       Alzheimers_Res_Ther  PMC3706879     GSE45534\n",
       "2       Alzheimers_Res_Ther  PMC3706879     GSE45534\n",
       "3       Alzheimers_Res_Ther  PMC3706879     GSE45534\n",
       "4       Alzheimers_Res_Ther  PMC3706879     GSE45534\n",
       "...                     ...         ...          ...\n",
       "144401          Dis_Markers  PMC3834650      GDS1059\n",
       "144402          Dis_Markers  PMC3834650      GDS1059\n",
       "144403          Dis_Markers  PMC3834650         GDS1\n",
       "144404     Case_Rep_Hematol  PMC6462343  PRJNA437812\n",
       "144405     Case_Rep_Hematol  PMC6462343    SRP135146\n",
       "\n",
       "[144406 rows x 3 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import data scraped from PubMed XML files\n",
    "pmcData = pd.read_csv('../data_tables/preFilterMatrix.csv')\n",
    "pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE48350</td>\n",
       "      <td>GSE48350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE36980</td>\n",
       "      <td>GSE36980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE12679</td>\n",
       "      <td>GSE12679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE12293</td>\n",
       "      <td>GSE12293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE46871</td>\n",
       "      <td>GSE46871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC5513091</td>\n",
       "      <td>GSE93678</td>\n",
       "      <td>GSE93678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>42</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC5819208</td>\n",
       "      <td>GSE44772</td>\n",
       "      <td>GSE44772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>43</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC5819208</td>\n",
       "      <td>GSE44771</td>\n",
       "      <td>GSE44771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>45</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC6015462</td>\n",
       "      <td>GSE60862</td>\n",
       "      <td>GSE60862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC6015462</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>47</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC6015462</td>\n",
       "      <td>GSE29378</td>\n",
       "      <td>GSE29378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>48</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC6015462</td>\n",
       "      <td>GSE36980</td>\n",
       "      <td>GSE36980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC6015462</td>\n",
       "      <td>GSE48350</td>\n",
       "      <td>GSE48350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>66</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC6015462</td>\n",
       "      <td>GSE9566</td>\n",
       "      <td>GSE9566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>77</td>\n",
       "      <td>Biol_Direct</td>\n",
       "      <td>PMC2928186</td>\n",
       "      <td>GSE2361</td>\n",
       "      <td>GSE2361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78</td>\n",
       "      <td>Biol_Direct</td>\n",
       "      <td>PMC2928186</td>\n",
       "      <td>GSE952</td>\n",
       "      <td>GSE952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>81</td>\n",
       "      <td>Biol_Direct</td>\n",
       "      <td>PMC1500996</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>GSE14;GSE15309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>82</td>\n",
       "      <td>Biol_Direct</td>\n",
       "      <td>PMC1634740</td>\n",
       "      <td>GDS422</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>83</td>\n",
       "      <td>Biol_Direct</td>\n",
       "      <td>PMC1634740</td>\n",
       "      <td>GDS596</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>84</td>\n",
       "      <td>Biol_Direct</td>\n",
       "      <td>PMC1634740</td>\n",
       "      <td>GDS181</td>\n",
       "      <td>GSE96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144347</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6348834</td>\n",
       "      <td>GSE65858</td>\n",
       "      <td>GSE65858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144350</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6341252</td>\n",
       "      <td>GSE16011</td>\n",
       "      <td>GSE16011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144354</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6387694</td>\n",
       "      <td>GSE53897</td>\n",
       "      <td>GSE53897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144358</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6387694</td>\n",
       "      <td>GPL6101</td>\n",
       "      <td>GSE63561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144361</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6476032</td>\n",
       "      <td>GSE19491</td>\n",
       "      <td>GSE19491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144362</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6476032</td>\n",
       "      <td>GSE37250</td>\n",
       "      <td>GSE37250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144363</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6476032</td>\n",
       "      <td>GSE42834</td>\n",
       "      <td>GSE42834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144373</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6481100</td>\n",
       "      <td>GSE10072</td>\n",
       "      <td>GSE10072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144374</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6481100</td>\n",
       "      <td>GSE32863</td>\n",
       "      <td>GSE32863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144375</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6481100</td>\n",
       "      <td>GSE85841</td>\n",
       "      <td>GSE85841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144377</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6481100</td>\n",
       "      <td>GPL96</td>\n",
       "      <td>GSE19301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144379</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6481100</td>\n",
       "      <td>GPL6884</td>\n",
       "      <td>GSE30682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144381</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6481100</td>\n",
       "      <td>GPL20115</td>\n",
       "      <td>GSE115018;GSE115019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144388</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6545767</td>\n",
       "      <td>GSE13597</td>\n",
       "      <td>GSE13597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144389</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6545767</td>\n",
       "      <td>GSE12452</td>\n",
       "      <td>GSE12452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144393</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6545767</td>\n",
       "      <td>GSE32323</td>\n",
       "      <td>GSE32323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144394</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6545767</td>\n",
       "      <td>GSE14407</td>\n",
       "      <td>GSE14407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144395</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589307</td>\n",
       "      <td>GSE73953</td>\n",
       "      <td>GSE73953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144396</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589307</td>\n",
       "      <td>GSE58539</td>\n",
       "      <td>GSE58539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144397</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589307</td>\n",
       "      <td>GSE14795</td>\n",
       "      <td>GSE14795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144400</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589301</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GSE62254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144401</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "      <td>GSE2191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144403</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144404</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "      <td>SRP135146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>144405</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRP135146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78620 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    journal      pmc_ID    accession  converted_accession\n",
       "0       Alzheimers_Res_Ther  PMC3707052         GSM1               GSE506\n",
       "1       Alzheimers_Res_Ther  PMC3706879     GSE45534             GSE45534\n",
       "5       Alzheimers_Res_Ther  PMC4255636      GSE5281              GSE5281\n",
       "6       Alzheimers_Res_Ther  PMC4731966     GSE67036             GSE67036\n",
       "7       Alzheimers_Res_Ther  PMC4731966      GSE1297              GSE1297\n",
       "8       Alzheimers_Res_Ther  PMC4731966     GSE48350             GSE48350\n",
       "9       Alzheimers_Res_Ther  PMC4731966     GSE36980             GSE36980\n",
       "10      Alzheimers_Res_Ther  PMC4731966     GSE12679             GSE12679\n",
       "11      Alzheimers_Res_Ther  PMC4731966     GSE12293             GSE12293\n",
       "17      Alzheimers_Res_Ther  PMC4731966     GSE46871             GSE46871\n",
       "41      Alzheimers_Res_Ther  PMC5513091     GSE93678             GSE93678\n",
       "42      Alzheimers_Res_Ther  PMC5819208     GSE44772             GSE44772\n",
       "43      Alzheimers_Res_Ther  PMC5819208     GSE44771             GSE44771\n",
       "45      Alzheimers_Res_Ther  PMC6015462     GSE60862             GSE60862\n",
       "46      Alzheimers_Res_Ther  PMC6015462      GSE5281              GSE5281\n",
       "47      Alzheimers_Res_Ther  PMC6015462     GSE29378             GSE29378\n",
       "48      Alzheimers_Res_Ther  PMC6015462     GSE36980             GSE36980\n",
       "49      Alzheimers_Res_Ther  PMC6015462     GSE48350             GSE48350\n",
       "66      Alzheimers_Res_Ther  PMC6015462      GSE9566              GSE9566\n",
       "77              Biol_Direct  PMC2928186      GSE2361              GSE2361\n",
       "78              Biol_Direct  PMC2928186       GSE952               GSE952\n",
       "81              Biol_Direct  PMC1500996         GPL4       GSE14;GSE15309\n",
       "82              Biol_Direct  PMC1634740       GDS422                  NaN\n",
       "83              Biol_Direct  PMC1634740       GDS596                  NaN\n",
       "84              Biol_Direct  PMC1634740       GDS181                GSE96\n",
       "...                     ...         ...          ...                  ...\n",
       "144347          Dis_Markers  PMC6348834     GSE65858             GSE65858\n",
       "144350          Dis_Markers  PMC6341252     GSE16011             GSE16011\n",
       "144354          Dis_Markers  PMC6387694     GSE53897             GSE53897\n",
       "144358          Dis_Markers  PMC6387694      GPL6101             GSE63561\n",
       "144361          Dis_Markers  PMC6476032     GSE19491             GSE19491\n",
       "144362          Dis_Markers  PMC6476032     GSE37250             GSE37250\n",
       "144363          Dis_Markers  PMC6476032     GSE42834             GSE42834\n",
       "144373          Dis_Markers  PMC6481100     GSE10072             GSE10072\n",
       "144374          Dis_Markers  PMC6481100     GSE32863             GSE32863\n",
       "144375          Dis_Markers  PMC6481100     GSE85841             GSE85841\n",
       "144377          Dis_Markers  PMC6481100        GPL96             GSE19301\n",
       "144379          Dis_Markers  PMC6481100      GPL6884             GSE30682\n",
       "144381          Dis_Markers  PMC6481100     GPL20115  GSE115018;GSE115019\n",
       "144388          Dis_Markers  PMC6545767     GSE13597             GSE13597\n",
       "144389          Dis_Markers  PMC6545767     GSE12452             GSE12452\n",
       "144393          Dis_Markers  PMC6545767     GSE32323             GSE32323\n",
       "144394          Dis_Markers  PMC6545767     GSE14407             GSE14407\n",
       "144395          Dis_Markers  PMC6589307     GSE73953             GSE73953\n",
       "144396          Dis_Markers  PMC6589307     GSE58539             GSE58539\n",
       "144397          Dis_Markers  PMC6589307     GSE14795             GSE14795\n",
       "144400          Dis_Markers  PMC6589301     GSE62254             GSE62254\n",
       "144401          Dis_Markers  PMC3834650      GDS1059              GSE2191\n",
       "144403          Dis_Markers  PMC3834650         GDS1                  NaN\n",
       "144404     Case_Rep_Hematol  PMC6462343  PRJNA437812            SRP135146\n",
       "144405     Case_Rep_Hematol  PMC6462343    SRP135146            SRP135146\n",
       "\n",
       "[78620 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check every accession against the corresponding reference, and convert to STUDY/SERIES\n",
    "# add converted accessions to a new list, to be added as a column later\n",
    "# CAUTION: long input list, loop takes several hours to run\n",
    "\n",
    "converted_acc = []\n",
    "\n",
    "for acc in pmcData['accession']:\n",
    "    style = ''\n",
    "    # assign a style to the accession, corresponds to column names\n",
    "    # SRA styles\n",
    "    if('SRP' in acc or 'ERP' in acc or 'DRP' in acc):\n",
    "        style = 'SRAStudy'\n",
    "    elif('SRR' in acc or 'ERR' in acc or 'DRR' in acc):\n",
    "        style = 'Run'\n",
    "    elif('SRX' in acc or 'ERX' in acc or 'DRX' in acc):\n",
    "        style = 'Experiment'\n",
    "    elif('PRJNA' in acc or 'PRJD' in acc or 'PRJEB' in acc):\n",
    "        style = 'BioProject'\n",
    "    elif('SRA' in acc or 'ERA' in acc or 'DRA' in acc):\n",
    "        style = 'Submission'\n",
    "    elif('SRS' in acc or 'ERS' in acc or 'DRS' in acc):\n",
    "        style = 'Sample'\n",
    "    # GEO styles\n",
    "    elif('GSE' in acc):\n",
    "        style = 'Series'\n",
    "    elif('GSM' in acc):\n",
    "        style = 'Accession'\n",
    "    elif('GPL' in acc):\n",
    "        style = 'Platform'\n",
    "    elif('GDS' in acc):\n",
    "        style = 'Datasets'\n",
    "    else:\n",
    "        style = ''\n",
    "        \n",
    "    # for SRA accessions, check if they exist in the SRA reference\n",
    "    # ...and add corresponding STUDY ID to new column\n",
    "    if(style == 'SRAStudy' or style == 'Run' or style == 'Experiment' or \n",
    "       style == 'BioProject' or style == 'Submission' or style == 'Sample'):\n",
    "        if(not sraChecksOut(acc, style)):\n",
    "            converted_acc.append('NaN')\n",
    "        else:\n",
    "            converted_acc.append(grabRelated(sraReference, acc, style, 'SRAStudy'))\n",
    "    # for GEO accessions, check if they exist in the GEO reference\n",
    "    # ...and add corresponding SERIES ID to new column\n",
    "    elif(style == 'Series' or style == 'Accession' or style == 'Platform' or style == 'Datasets'):\n",
    "        if(not geoChecksOut(acc, style)):\n",
    "            converted_acc.append('NaN')\n",
    "        else:\n",
    "            converted_acc.append(grabRelated(geoReference, acc, style, 'Series'))\n",
    "    # something doesn't belong, record as missing\n",
    "    else:\n",
    "        converted_acc.append('NaN')\n",
    "\n",
    "# add the converted accession list as a new column in PubMed data table\n",
    "# pd.set_option('display.min_rows', 50)\n",
    "pmcData.loc[:,'converted_accession'] = converted_acc\n",
    "pmcData = pmcData.drop_duplicates()\n",
    "pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save 'checkpoint' to CSV file\n",
    "# pmcData.to_csv('postFilterMatrix.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78615</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589301</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GSE62254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78616</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "      <td>GSE2191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78617</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78618</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "      <td>SRP135146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78619</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRP135146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78620 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   journal      pmc_ID    accession converted_accession\n",
       "0      Alzheimers_Res_Ther  PMC3707052         GSM1              GSE506\n",
       "1      Alzheimers_Res_Ther  PMC3706879     GSE45534            GSE45534\n",
       "2      Alzheimers_Res_Ther  PMC4255636      GSE5281             GSE5281\n",
       "3      Alzheimers_Res_Ther  PMC4731966     GSE67036            GSE67036\n",
       "4      Alzheimers_Res_Ther  PMC4731966      GSE1297             GSE1297\n",
       "...                    ...         ...          ...                 ...\n",
       "78615          Dis_Markers  PMC6589301     GSE62254            GSE62254\n",
       "78616          Dis_Markers  PMC3834650      GDS1059             GSE2191\n",
       "78617          Dis_Markers  PMC3834650         GDS1                 NaN\n",
       "78618     Case_Rep_Hematol  PMC6462343  PRJNA437812           SRP135146\n",
       "78619     Case_Rep_Hematol  PMC6462343    SRP135146           SRP135146\n",
       "\n",
       "[78620 rows x 4 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# temporary cell saving time from big loop...\n",
    "\n",
    "pmcData = pd.read_csv('../data_tables/postFilterMatrix.csv')\n",
    "pmcData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step II: Add all desired factors about each SRA or GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "      <th>repository</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78615</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589301</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78616</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "      <td>GSE2191</td>\n",
       "      <td>GEO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78617</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78618</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78619</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78620 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   journal      pmc_ID    accession converted_accession  \\\n",
       "0      Alzheimers_Res_Ther  PMC3707052         GSM1              GSE506   \n",
       "1      Alzheimers_Res_Ther  PMC3706879     GSE45534            GSE45534   \n",
       "2      Alzheimers_Res_Ther  PMC4255636      GSE5281             GSE5281   \n",
       "3      Alzheimers_Res_Ther  PMC4731966     GSE67036            GSE67036   \n",
       "4      Alzheimers_Res_Ther  PMC4731966      GSE1297             GSE1297   \n",
       "...                    ...         ...          ...                 ...   \n",
       "78615          Dis_Markers  PMC6589301     GSE62254            GSE62254   \n",
       "78616          Dis_Markers  PMC3834650      GDS1059             GSE2191   \n",
       "78617          Dis_Markers  PMC3834650         GDS1                 NaN   \n",
       "78618     Case_Rep_Hematol  PMC6462343  PRJNA437812           SRP135146   \n",
       "78619     Case_Rep_Hematol  PMC6462343    SRP135146           SRP135146   \n",
       "\n",
       "      repository  \n",
       "0            GEO  \n",
       "1            GEO  \n",
       "2            GEO  \n",
       "3            GEO  \n",
       "4            GEO  \n",
       "...          ...  \n",
       "78615        GEO  \n",
       "78616        GEO  \n",
       "78617        NaN  \n",
       "78618        SRA  \n",
       "78619        SRA  \n",
       "\n",
       "[78620 rows x 5 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column tagging each accession as GEO or SRA\n",
    "\n",
    "repoList = []\n",
    "\n",
    "for i in pmcData['converted_accession']:\n",
    "    if(type(i) == str):\n",
    "        if('GSE' in i):\n",
    "            repoList.append('GEO')\n",
    "        elif('SRP' in i or 'ERP' in i or 'DRP' in i):\n",
    "            repoList.append('SRA')\n",
    "        else:\n",
    "            repoList.append('NaN')\n",
    "    else:\n",
    "        repoList.append('NaN')\n",
    "        \n",
    "pmcData['repository'] = repoList\n",
    "pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>spots</th>\n",
       "      <th>bases</th>\n",
       "      <th>spots_with_mates</th>\n",
       "      <th>avgLength</th>\n",
       "      <th>size_MB</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>download_path</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>LibraryName</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>InsertSize</th>\n",
       "      <th>InsertDev</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SRAStudy</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Study_Pubmed_id</th>\n",
       "      <th>ProjectID</th>\n",
       "      <th>Sample</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>SampleType</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>SampleName</th>\n",
       "      <th>g1k_pop_code</th>\n",
       "      <th>source</th>\n",
       "      <th>g1k_analysis_group</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Disease</th>\n",
       "      <th>Tumor</th>\n",
       "      <th>Affection_Status</th>\n",
       "      <th>Analyte_Type</th>\n",
       "      <th>Histological_Type</th>\n",
       "      <th>Body_Site</th>\n",
       "      <th>CenterName</th>\n",
       "      <th>Submission</th>\n",
       "      <th>dbgap_study_accession</th>\n",
       "      <th>Consent</th>\n",
       "      <th>RunHash</th>\n",
       "      <th>ReadHash</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ERR3550121</td>\n",
       "      <td>2019-09-25 19:35:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX3556726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>204335</td>\n",
       "      <td>ERS3773255</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simple</td>\n",
       "      <td>0</td>\n",
       "      <td>Leishmania infantum</td>\n",
       "      <td>0c4ac31a-dec5-11e9-ba7b-68b599768938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UOC-CCG</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ERR3550142</td>\n",
       "      <td>2019-09-25 19:35:41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX3556747</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>204335</td>\n",
       "      <td>ERS3773255</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simple</td>\n",
       "      <td>0</td>\n",
       "      <td>Leishmania infantum</td>\n",
       "      <td>0c4ac31a-dec5-11e9-ba7b-68b599768938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UOC-CCG</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ERR3550147</td>\n",
       "      <td>2019-09-25 19:35:41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX3556752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>204335</td>\n",
       "      <td>ERS3773263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simple</td>\n",
       "      <td>0</td>\n",
       "      <td>Leishmania donovani</td>\n",
       "      <td>0e2d3a2c-dec6-11e9-a5d1-68b59976a384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UOC-CCG</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ERR3550125</td>\n",
       "      <td>2019-09-25 19:35:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX3556730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>204335</td>\n",
       "      <td>ERS3773263</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simple</td>\n",
       "      <td>0</td>\n",
       "      <td>Leishmania donovani</td>\n",
       "      <td>0e2d3a2c-dec6-11e9-a5d1-68b59976a384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UOC-CCG</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ERR3550137</td>\n",
       "      <td>2019-09-25 19:35:40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX3556742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>250</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ERP000767</td>\n",
       "      <td>PRJEB2600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>204335</td>\n",
       "      <td>ERS3773252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>simple</td>\n",
       "      <td>0</td>\n",
       "      <td>Leishmania donovani</td>\n",
       "      <td>188c9a6e-deba-11e9-ba7b-68b599768938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UOC-CCG</td>\n",
       "      <td>ERA2143351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676058</td>\n",
       "      <td>SRR049674</td>\n",
       "      <td>2010-05-27 09:26:35</td>\n",
       "      <td>2014-05-27 02:57:13</td>\n",
       "      <td>35888370</td>\n",
       "      <td>1794418500</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>1504</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...</td>\n",
       "      <td>SRX020916</td>\n",
       "      <td>VVWTA 0h A</td>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ABI_SOLID</td>\n",
       "      <td>AB SOLiD System 3.0</td>\n",
       "      <td>SRP002493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS073299</td>\n",
       "      <td>SAMN00013848</td>\n",
       "      <td>simple</td>\n",
       "      <td>10245</td>\n",
       "      <td>Vaccinia virus</td>\n",
       "      <td>VACV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NIAID-RML-RTS</td>\n",
       "      <td>SRA017695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>1C9EE0CEFAE2352A0321B24EAB64EE18</td>\n",
       "      <td>753ED95CDA0B112FD877F38960613605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676059</td>\n",
       "      <td>SRR049675</td>\n",
       "      <td>2010-05-27 09:26:35</td>\n",
       "      <td>2014-05-27 02:58:17</td>\n",
       "      <td>40264881</td>\n",
       "      <td>2013244050</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>1696</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...</td>\n",
       "      <td>SRX021028</td>\n",
       "      <td>VVWTA 0.5A</td>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ABI_SOLID</td>\n",
       "      <td>AB SOLiD System 3.0</td>\n",
       "      <td>SRP002493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS073299</td>\n",
       "      <td>SAMN00013848</td>\n",
       "      <td>simple</td>\n",
       "      <td>10245</td>\n",
       "      <td>Vaccinia virus</td>\n",
       "      <td>VACV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NIAID-RML-RTS</td>\n",
       "      <td>SRA017695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>9F96F35BE762574A084985FD9C918C28</td>\n",
       "      <td>392587111A738BC44AE8878E649D052A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676060</td>\n",
       "      <td>SRR049677</td>\n",
       "      <td>2010-05-27 09:26:35</td>\n",
       "      <td>2014-05-27 02:58:58</td>\n",
       "      <td>40852259</td>\n",
       "      <td>2042612950</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>1732</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...</td>\n",
       "      <td>SRX021029</td>\n",
       "      <td>VVWTA 1h A</td>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ABI_SOLID</td>\n",
       "      <td>AB SOLiD System 3.0</td>\n",
       "      <td>SRP002493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS073299</td>\n",
       "      <td>SAMN00013848</td>\n",
       "      <td>simple</td>\n",
       "      <td>10245</td>\n",
       "      <td>Vaccinia virus</td>\n",
       "      <td>VACV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NIAID-RML-RTS</td>\n",
       "      <td>SRA017695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>22DCD0A93DAA207187EF6C6F6A835A55</td>\n",
       "      <td>416EBCA2B1A6AB70E4F778F16C36FCE5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676061</td>\n",
       "      <td>SRR042496</td>\n",
       "      <td>2010-05-11 15:16:41</td>\n",
       "      <td>2012-01-19 16:52:13</td>\n",
       "      <td>47760</td>\n",
       "      <td>16505980</td>\n",
       "      <td>0</td>\n",
       "      <td>345</td>\n",
       "      <td>42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...</td>\n",
       "      <td>SRX020084</td>\n",
       "      <td>VirRnaMgMosqCeI</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LS454</td>\n",
       "      <td>454 GS FLX Titanium</td>\n",
       "      <td>SRP002064</td>\n",
       "      <td>PRJNA46359</td>\n",
       "      <td>2.0</td>\n",
       "      <td>46359</td>\n",
       "      <td>SRS011854</td>\n",
       "      <td>SAMN00012272</td>\n",
       "      <td>simple</td>\n",
       "      <td>1284618</td>\n",
       "      <td>environmental samples</td>\n",
       "      <td>VirRnaMgMosqCeI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAN DIEGO STATE UNIVERSITY</td>\n",
       "      <td>SRA012164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>633E95BCDFC18AD9C8B5DEAAFAE4DD3F</td>\n",
       "      <td>00AB1B11657DAE26FFECDAD623148A3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7676062</td>\n",
       "      <td>SRR043092</td>\n",
       "      <td>2010-05-10 17:56:13</td>\n",
       "      <td>2012-01-19 16:52:35</td>\n",
       "      <td>162876</td>\n",
       "      <td>43696870</td>\n",
       "      <td>0</td>\n",
       "      <td>268</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...</td>\n",
       "      <td>SRX020242</td>\n",
       "      <td>All HCV</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>PCR</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LS454</td>\n",
       "      <td>454 GS FLX</td>\n",
       "      <td>SRP002438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>SRS072567</td>\n",
       "      <td>SAMN00012789</td>\n",
       "      <td>simple</td>\n",
       "      <td>11103</td>\n",
       "      <td>Hepacivirus C</td>\n",
       "      <td>all_hcv</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNIVERSITY OF PENNSYLVANIA</td>\n",
       "      <td>SRA012580</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>1D70AED9981C7955082992A1AA787246</td>\n",
       "      <td>7C03FE9A87DE427C72136E8C1CA1781C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7676063 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Run          ReleaseDate             LoadDate     spots  \\\n",
       "0        ERR3550121  2019-09-25 19:35:40                  NaN         0   \n",
       "1        ERR3550142  2019-09-25 19:35:41                  NaN         0   \n",
       "2        ERR3550147  2019-09-25 19:35:41                  NaN         0   \n",
       "3        ERR3550125  2019-09-25 19:35:40                  NaN         0   \n",
       "4        ERR3550137  2019-09-25 19:35:40                  NaN         0   \n",
       "...             ...                  ...                  ...       ...   \n",
       "7676058   SRR049674  2010-05-27 09:26:35  2014-05-27 02:57:13  35888370   \n",
       "7676059   SRR049675  2010-05-27 09:26:35  2014-05-27 02:58:17  40264881   \n",
       "7676060   SRR049677  2010-05-27 09:26:35  2014-05-27 02:58:58  40852259   \n",
       "7676061   SRR042496  2010-05-11 15:16:41  2012-01-19 16:52:13     47760   \n",
       "7676062   SRR043092  2010-05-10 17:56:13  2012-01-19 16:52:35    162876   \n",
       "\n",
       "              bases  spots_with_mates  avgLength  size_MB AssemblyName  \\\n",
       "0                 0                 0          0        0          NaN   \n",
       "1                 0                 0          0        0          NaN   \n",
       "2                 0                 0          0        0          NaN   \n",
       "3                 0                 0          0        0          NaN   \n",
       "4                 0                 0          0        0          NaN   \n",
       "...             ...               ...        ...      ...          ...   \n",
       "7676058  1794418500                 0         50     1504          NaN   \n",
       "7676059  2013244050                 0         50     1696          NaN   \n",
       "7676060  2042612950                 0         50     1732          NaN   \n",
       "7676061    16505980                 0        345       42          NaN   \n",
       "7676062    43696870                 0        268       80          NaN   \n",
       "\n",
       "                                             download_path  Experiment  \\\n",
       "0                                                      NaN  ERX3556726   \n",
       "1                                                      NaN  ERX3556747   \n",
       "2                                                      NaN  ERX3556752   \n",
       "3                                                      NaN  ERX3556730   \n",
       "4                                                      NaN  ERX3556742   \n",
       "...                                                    ...         ...   \n",
       "7676058  https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...   SRX020916   \n",
       "7676059  https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...   SRX021028   \n",
       "7676060  https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...   SRX021029   \n",
       "7676061  https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...   SRX020084   \n",
       "7676062  https://sra-downloadb.be-md.ncbi.nlm.nih.gov/s...   SRX020242   \n",
       "\n",
       "             LibraryName LibraryStrategy LibrarySelection LibrarySource  \\\n",
       "0                    NaN             WGS           RANDOM       GENOMIC   \n",
       "1                    NaN             WGS           RANDOM       GENOMIC   \n",
       "2                    NaN             WGS           RANDOM       GENOMIC   \n",
       "3                    NaN             WGS           RANDOM       GENOMIC   \n",
       "4                    NaN             WGS           RANDOM       GENOMIC   \n",
       "...                  ...             ...              ...           ...   \n",
       "7676058       VVWTA 0h A         FL-cDNA           RT-PCR     VIRAL RNA   \n",
       "7676059       VVWTA 0.5A         FL-cDNA           RT-PCR     VIRAL RNA   \n",
       "7676060       VVWTA 1h A         FL-cDNA           RT-PCR     VIRAL RNA   \n",
       "7676061  VirRnaMgMosqCeI           OTHER       RANDOM PCR     VIRAL RNA   \n",
       "7676062          All HCV        AMPLICON              PCR     VIRAL RNA   \n",
       "\n",
       "        LibraryLayout  InsertSize  InsertDev   Platform                Model  \\\n",
       "0              PAIRED         250        0.0   ILLUMINA  Illumina HiSeq 2000   \n",
       "1              PAIRED         250        0.0   ILLUMINA  Illumina HiSeq 2000   \n",
       "2              PAIRED         250        0.0   ILLUMINA  Illumina HiSeq 2000   \n",
       "3              PAIRED         250        0.0   ILLUMINA  Illumina HiSeq 2000   \n",
       "4              PAIRED         250        0.0   ILLUMINA  Illumina HiSeq 2000   \n",
       "...               ...         ...        ...        ...                  ...   \n",
       "7676058        SINGLE           0        0.0  ABI_SOLID  AB SOLiD System 3.0   \n",
       "7676059        SINGLE           0        0.0  ABI_SOLID  AB SOLiD System 3.0   \n",
       "7676060        SINGLE           0        0.0  ABI_SOLID  AB SOLiD System 3.0   \n",
       "7676061        SINGLE           0        0.0      LS454  454 GS FLX Titanium   \n",
       "7676062        SINGLE           0        0.0      LS454           454 GS FLX   \n",
       "\n",
       "          SRAStudy  BioProject Study_Pubmed_id  ProjectID      Sample  \\\n",
       "0        ERP000767   PRJEB2600             NaN     204335  ERS3773255   \n",
       "1        ERP000767   PRJEB2600             NaN     204335  ERS3773255   \n",
       "2        ERP000767   PRJEB2600             NaN     204335  ERS3773263   \n",
       "3        ERP000767   PRJEB2600             NaN     204335  ERS3773263   \n",
       "4        ERP000767   PRJEB2600             NaN     204335  ERS3773252   \n",
       "...            ...         ...             ...        ...         ...   \n",
       "7676058  SRP002493         NaN             2.0          0   SRS073299   \n",
       "7676059  SRP002493         NaN             2.0          0   SRS073299   \n",
       "7676060  SRP002493         NaN             2.0          0   SRS073299   \n",
       "7676061  SRP002064  PRJNA46359             2.0      46359   SRS011854   \n",
       "7676062  SRP002438         NaN             2.0          0   SRS072567   \n",
       "\n",
       "            BioSample SampleType    TaxID         ScientificName  \\\n",
       "0                 NaN     simple        0    Leishmania infantum   \n",
       "1                 NaN     simple        0    Leishmania infantum   \n",
       "2                 NaN     simple        0    Leishmania donovani   \n",
       "3                 NaN     simple        0    Leishmania donovani   \n",
       "4                 NaN     simple        0    Leishmania donovani   \n",
       "...               ...        ...      ...                    ...   \n",
       "7676058  SAMN00013848     simple    10245         Vaccinia virus   \n",
       "7676059  SAMN00013848     simple    10245         Vaccinia virus   \n",
       "7676060  SAMN00013848     simple    10245         Vaccinia virus   \n",
       "7676061  SAMN00012272     simple  1284618  environmental samples   \n",
       "7676062  SAMN00012789     simple    11103          Hepacivirus C   \n",
       "\n",
       "                                   SampleName g1k_pop_code source  \\\n",
       "0        0c4ac31a-dec5-11e9-ba7b-68b599768938          NaN    NaN   \n",
       "1        0c4ac31a-dec5-11e9-ba7b-68b599768938          NaN    NaN   \n",
       "2        0e2d3a2c-dec6-11e9-a5d1-68b59976a384          NaN    NaN   \n",
       "3        0e2d3a2c-dec6-11e9-a5d1-68b59976a384          NaN    NaN   \n",
       "4        188c9a6e-deba-11e9-ba7b-68b599768938          NaN    NaN   \n",
       "...                                       ...          ...    ...   \n",
       "7676058                                  VACV          NaN    NaN   \n",
       "7676059                                  VACV          NaN    NaN   \n",
       "7676060                                  VACV          NaN    NaN   \n",
       "7676061                       VirRnaMgMosqCeI          NaN    NaN   \n",
       "7676062                               all_hcv          NaN    NaN   \n",
       "\n",
       "        g1k_analysis_group Subject_ID  Sex Disease Tumor Affection_Status  \\\n",
       "0                      NaN        NaN  NaN     NaN    no              NaN   \n",
       "1                      NaN        NaN  NaN     NaN    no              NaN   \n",
       "2                      NaN        NaN  NaN     NaN    no              NaN   \n",
       "3                      NaN        NaN  NaN     NaN    no              NaN   \n",
       "4                      NaN        NaN  NaN     NaN    no              NaN   \n",
       "...                    ...        ...  ...     ...   ...              ...   \n",
       "7676058                NaN        NaN  NaN     NaN    no              NaN   \n",
       "7676059                NaN        NaN  NaN     NaN    no              NaN   \n",
       "7676060                NaN        NaN  NaN     NaN    no              NaN   \n",
       "7676061                NaN        NaN  NaN     NaN    no              NaN   \n",
       "7676062                NaN        NaN  NaN     NaN    no              NaN   \n",
       "\n",
       "        Analyte_Type Histological_Type Body_Site                  CenterName  \\\n",
       "0                NaN               NaN       NaN                     UOC-CCG   \n",
       "1                NaN               NaN       NaN                     UOC-CCG   \n",
       "2                NaN               NaN       NaN                     UOC-CCG   \n",
       "3                NaN               NaN       NaN                     UOC-CCG   \n",
       "4                NaN               NaN       NaN                     UOC-CCG   \n",
       "...              ...               ...       ...                         ...   \n",
       "7676058          NaN               NaN       NaN               NIAID-RML-RTS   \n",
       "7676059          NaN               NaN       NaN               NIAID-RML-RTS   \n",
       "7676060          NaN               NaN       NaN               NIAID-RML-RTS   \n",
       "7676061          NaN               NaN       NaN  SAN DIEGO STATE UNIVERSITY   \n",
       "7676062          NaN               NaN       NaN  UNIVERSITY OF PENNSYLVANIA   \n",
       "\n",
       "         Submission dbgap_study_accession Consent  \\\n",
       "0        ERA2143351                   NaN  public   \n",
       "1        ERA2143351                   NaN  public   \n",
       "2        ERA2143351                   NaN  public   \n",
       "3        ERA2143351                   NaN  public   \n",
       "4        ERA2143351                   NaN  public   \n",
       "...             ...                   ...     ...   \n",
       "7676058   SRA017695                   NaN  public   \n",
       "7676059   SRA017695                   NaN  public   \n",
       "7676060   SRA017695                   NaN  public   \n",
       "7676061   SRA012164                   NaN  public   \n",
       "7676062   SRA012580                   NaN  public   \n",
       "\n",
       "                                  RunHash                          ReadHash  \n",
       "0                                     NaN                               NaN  \n",
       "1                                     NaN                               NaN  \n",
       "2                                     NaN                               NaN  \n",
       "3                                     NaN                               NaN  \n",
       "4                                     NaN                               NaN  \n",
       "...                                   ...                               ...  \n",
       "7676058  1C9EE0CEFAE2352A0321B24EAB64EE18  753ED95CDA0B112FD877F38960613605  \n",
       "7676059  9F96F35BE762574A084985FD9C918C28  392587111A738BC44AE8878E649D052A  \n",
       "7676060  22DCD0A93DAA207187EF6C6F6A835A55  416EBCA2B1A6AB70E4F778F16C36FCE5  \n",
       "7676061  633E95BCDFC18AD9C8B5DEAAFAE4DD3F  00AB1B11657DAE26FFECDAD623148A3C  \n",
       "7676062  1D70AED9981C7955082992A1AA787246  7C03FE9A87DE427C72136E8C1CA1781C  \n",
       "\n",
       "[7676063 rows x 47 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import SRA attribute data\n",
    "# CAUTION: large file! Time delay on import...\n",
    "\n",
    "sraAttributes = pd.read_csv('../sra_complete_runs.csv', error_bad_lines = False, low_memory=False)\n",
    "pd.set_option('display.max_columns', 50)\n",
    "sraAttributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Platform</th>\n",
       "      <th>Title</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Data Rows</th>\n",
       "      <th>Samples Count</th>\n",
       "      <th>Series Count</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "      <th>Series</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Datasets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>SAGE:10:NlaIII:Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>265577</td>\n",
       "      <td>580</td>\n",
       "      <td>58</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>SAGE:10:NlaIII:Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>265577</td>\n",
       "      <td>580</td>\n",
       "      <td>58</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GSM2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>SAGE:10:NlaIII:Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>265577</td>\n",
       "      <td>580</td>\n",
       "      <td>58</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>GSE10</td>\n",
       "      <td>GSM571</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>SAGE:10:NlaIII:Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>265577</td>\n",
       "      <td>580</td>\n",
       "      <td>58</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>GSE10</td>\n",
       "      <td>GSM572</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>GPL4</td>\n",
       "      <td>SAGE:10:NlaIII:Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>265577</td>\n",
       "      <td>580</td>\n",
       "      <td>58</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>GSE10</td>\n",
       "      <td>GSM573</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2961228</td>\n",
       "      <td>GPL27549</td>\n",
       "      <td>Illumina HiSeq 2000 (Escherichia coli str. K-1...</td>\n",
       "      <td>high-throughput sequencing</td>\n",
       "      <td>Escherichia coli str. K-12 substr. DH10B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 30, 2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2961229</td>\n",
       "      <td>GPL27550</td>\n",
       "      <td>Illumina HiSeq 4000 (Solanum chacoense)</td>\n",
       "      <td>high-throughput sequencing</td>\n",
       "      <td>Solanum chacoense</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 30, 2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2961230</td>\n",
       "      <td>GPL27552</td>\n",
       "      <td>Illumina HiSeq 2500 (Caenorhabditis elegans ,V...</td>\n",
       "      <td>high-throughput sequencing</td>\n",
       "      <td>Caenorhabditis elegans ,Vibrio parahaemolyticus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 30, 2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2961231</td>\n",
       "      <td>GPL27554</td>\n",
       "      <td>Illumina HiSeq 2000 (Trichoderma atroviride)</td>\n",
       "      <td>high-throughput sequencing</td>\n",
       "      <td>Trichoderma atroviride</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 30, 2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2961232</td>\n",
       "      <td>GPL27555</td>\n",
       "      <td>Illumina HiSeq 2500 (Caenorhabditis elegans; V...</td>\n",
       "      <td>high-throughput sequencing</td>\n",
       "      <td>Caenorhabditis elegans;Vibrio parahaemolyticus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Sep 30, 2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2961233 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Platform                                              Title  \\\n",
       "0            GPL4                        SAGE:10:NlaIII:Homo sapiens   \n",
       "1            GPL4                        SAGE:10:NlaIII:Homo sapiens   \n",
       "2            GPL4                        SAGE:10:NlaIII:Homo sapiens   \n",
       "3            GPL4                        SAGE:10:NlaIII:Homo sapiens   \n",
       "4            GPL4                        SAGE:10:NlaIII:Homo sapiens   \n",
       "...           ...                                                ...   \n",
       "2961228  GPL27549  Illumina HiSeq 2000 (Escherichia coli str. K-1...   \n",
       "2961229  GPL27550            Illumina HiSeq 4000 (Solanum chacoense)   \n",
       "2961230  GPL27552  Illumina HiSeq 2500 (Caenorhabditis elegans ,V...   \n",
       "2961231  GPL27554       Illumina HiSeq 2000 (Trichoderma atroviride)   \n",
       "2961232  GPL27555  Illumina HiSeq 2500 (Caenorhabditis elegans; V...   \n",
       "\n",
       "                         Technology  \\\n",
       "0                       SAGE NlaIII   \n",
       "1                       SAGE NlaIII   \n",
       "2                       SAGE NlaIII   \n",
       "3                       SAGE NlaIII   \n",
       "4                       SAGE NlaIII   \n",
       "...                             ...   \n",
       "2961228  high-throughput sequencing   \n",
       "2961229  high-throughput sequencing   \n",
       "2961230  high-throughput sequencing   \n",
       "2961231  high-throughput sequencing   \n",
       "2961232  high-throughput sequencing   \n",
       "\n",
       "                                                Taxonomy  Data Rows  \\\n",
       "0                                           Homo sapiens     265577   \n",
       "1                                           Homo sapiens     265577   \n",
       "2                                           Homo sapiens     265577   \n",
       "3                                           Homo sapiens     265577   \n",
       "4                                           Homo sapiens     265577   \n",
       "...                                                  ...        ...   \n",
       "2961228         Escherichia coli str. K-12 substr. DH10B          0   \n",
       "2961229                                Solanum chacoense          0   \n",
       "2961230  Caenorhabditis elegans ,Vibrio parahaemolyticus          0   \n",
       "2961231                           Trichoderma atroviride          0   \n",
       "2961232   Caenorhabditis elegans;Vibrio parahaemolyticus          0   \n",
       "\n",
       "         Samples Count  Series Count Contact  Release Date  Series Accession  \\\n",
       "0                  580            58     GEO  Sep 28, 2000  GSE506      GSM1   \n",
       "1                  580            58     GEO  Sep 28, 2000  GSE506      GSM2   \n",
       "2                  580            58     GEO  Sep 28, 2000   GSE10    GSM571   \n",
       "3                  580            58     GEO  Sep 28, 2000   GSE10    GSM572   \n",
       "4                  580            58     GEO  Sep 28, 2000   GSE10    GSM573   \n",
       "...                ...           ...     ...           ...     ...       ...   \n",
       "2961228              0             0     GEO  Sep 30, 2019     NaN       NaN   \n",
       "2961229              0             0     GEO  Sep 30, 2019     NaN       NaN   \n",
       "2961230              0             0     GEO  Sep 30, 2019     NaN       NaN   \n",
       "2961231              0             0     GEO  Sep 30, 2019     NaN       NaN   \n",
       "2961232              0             0     GEO  Sep 30, 2019     NaN       NaN   \n",
       "\n",
       "        Datasets  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "...          ...  \n",
       "2961228      NaN  \n",
       "2961229      NaN  \n",
       "2961230      NaN  \n",
       "2961231      NaN  \n",
       "2961232      NaN  \n",
       "\n",
       "[2961233 rows x 12 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import GEO attribute data and add Series column\n",
    "\n",
    "geoAttributes = pd.read_csv('../geo_platforms_table.csv')\n",
    "geoAttributes.rename(columns={'Accession':'Platform'}, inplace=True)\n",
    "geoAttributes = pd.merge(geoAttributes, geoReference, how = 'left', on = 'Platform')\n",
    "geoAttributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.unique(geoAttributes['Technology'].tolist()).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "      <th>repository</th>\n",
       "      <th>pmc_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78615</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589301</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78616</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "      <td>GSE2191</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78617</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78618</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78619</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78620 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   journal      pmc_ID    accession converted_accession  \\\n",
       "0      Alzheimers_Res_Ther  PMC3707052         GSM1              GSE506   \n",
       "1      Alzheimers_Res_Ther  PMC3706879     GSE45534            GSE45534   \n",
       "2      Alzheimers_Res_Ther  PMC4255636      GSE5281             GSE5281   \n",
       "3      Alzheimers_Res_Ther  PMC4731966     GSE67036            GSE67036   \n",
       "4      Alzheimers_Res_Ther  PMC4731966      GSE1297             GSE1297   \n",
       "...                    ...         ...          ...                 ...   \n",
       "78615          Dis_Markers  PMC6589301     GSE62254            GSE62254   \n",
       "78616          Dis_Markers  PMC3834650      GDS1059             GSE2191   \n",
       "78617          Dis_Markers  PMC3834650         GDS1                 NaN   \n",
       "78618     Case_Rep_Hematol  PMC6462343  PRJNA437812           SRP135146   \n",
       "78619     Case_Rep_Hematol  PMC6462343    SRP135146           SRP135146   \n",
       "\n",
       "      repository  pmc_date  \n",
       "0            GEO      2013  \n",
       "1            GEO      2013  \n",
       "2            GEO      2014  \n",
       "3            GEO      2016  \n",
       "4            GEO      2016  \n",
       "...          ...       ...  \n",
       "78615        GEO      2019  \n",
       "78616        GEO      2013  \n",
       "78617        NaN      2013  \n",
       "78618        SRA      2019  \n",
       "78619        SRA      2019  \n",
       "\n",
       "[78620 rows x 6 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add column for paper publish date\n",
    "# this data was scraped from XML files on the hoffman2 cluster: /u/scratch/n/nikodm/pmcOA/\n",
    "\n",
    "pmc_dates = pd.read_csv('../data_lists/postFilterDates.txt')\n",
    "\n",
    "pmcData.loc[:,'pmc_date'] = pmc_dates['date'].tolist()\n",
    "pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get every factor we're interested in from our tables of GEO and SRA metadata...\n",
    "\n",
    "# take a slice of the GEO and SRA attribute tables with only the info we want\n",
    "# drop duplicate converted_accession entries to avoid unnecessarily large table\n",
    "slicedGEOAtt = geoAttributes[['Series', 'Release Date', 'Technology', 'Taxonomy']]\n",
    "slicedGEOAtt.columns = ['converted_accession', 'geoRelease', 'geoHardware', 'geoSpecies']\n",
    "slicedGEOAtt = slicedGEOAtt.drop_duplicates(subset = ['converted_accession'])\n",
    "\n",
    "slicedSRAAtt = sraAttributes[['SRAStudy', 'ReleaseDate', 'Model', \n",
    "                              'LibraryStrategy', 'ScientificName', \n",
    "                              'bases', 'avgLength', 'Consent']]\n",
    "slicedSRAAtt.columns = ['converted_accession', 'sraRelease', 'sraHardware', \n",
    "                        'sraLibrary_strategy', 'sraSpecies', \n",
    "                        'sraBases', 'sraAvg_length', 'sraAccess']\n",
    "slicedSRAAtt = slicedSRAAtt.drop_duplicates(subset = ['converted_accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge SRA attributes onto pmcData table\n",
    "mergedSRA = pd.merge(pmcData, slicedSRAAtt, how = 'left', on = 'converted_accession')\n",
    "mergedSRA = mergedSRA.drop_duplicates()\n",
    "\n",
    "# merge GEO attributes onto table of pmcData + SRA Attributes\n",
    "allFactors = pd.merge(mergedSRA, slicedGEOAtt, how = 'left', on = 'converted_accession')\n",
    "allFactors = allFactors.dropna(subset = ['converted_accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "      <th>repository</th>\n",
       "      <th>pmc_date</th>\n",
       "      <th>repository_date</th>\n",
       "      <th>species</th>\n",
       "      <th>hardware</th>\n",
       "      <th>sraLibrary_strategy</th>\n",
       "      <th>sraAvg_length</th>\n",
       "      <th>sraBases</th>\n",
       "      <th>sraAccess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "      <td>May 25, 2004</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2014</td>\n",
       "      <td>Nov 07, 2003</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016</td>\n",
       "      <td>May 06, 2013</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016</td>\n",
       "      <td>Mar 11, 2002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78614</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589307</td>\n",
       "      <td>GSE14795</td>\n",
       "      <td>GSE14795</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2019</td>\n",
       "      <td>Mar 11, 2002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78615</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589301</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78616</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "      <td>GSE2191</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "      <td>Mar 16, 2009</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78618</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019</td>\n",
       "      <td>2018-03-13 01:47:12</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>602.0</td>\n",
       "      <td>17743348.0</td>\n",
       "      <td>public</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78619</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019</td>\n",
       "      <td>2018-03-13 01:47:12</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>602.0</td>\n",
       "      <td>17743348.0</td>\n",
       "      <td>public</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73901 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   journal      pmc_ID    accession converted_accession  \\\n",
       "0      Alzheimers_Res_Ther  PMC3707052         GSM1              GSE506   \n",
       "1      Alzheimers_Res_Ther  PMC3706879     GSE45534            GSE45534   \n",
       "2      Alzheimers_Res_Ther  PMC4255636      GSE5281             GSE5281   \n",
       "3      Alzheimers_Res_Ther  PMC4731966     GSE67036            GSE67036   \n",
       "4      Alzheimers_Res_Ther  PMC4731966      GSE1297             GSE1297   \n",
       "...                    ...         ...          ...                 ...   \n",
       "78614          Dis_Markers  PMC6589307     GSE14795            GSE14795   \n",
       "78615          Dis_Markers  PMC6589301     GSE62254            GSE62254   \n",
       "78616          Dis_Markers  PMC3834650      GDS1059             GSE2191   \n",
       "78618     Case_Rep_Hematol  PMC6462343  PRJNA437812           SRP135146   \n",
       "78619     Case_Rep_Hematol  PMC6462343    SRP135146           SRP135146   \n",
       "\n",
       "      repository  pmc_date      repository_date               species  \\\n",
       "0            GEO      2013         Sep 28, 2000          Homo sapiens   \n",
       "1            GEO      2013         May 25, 2004          Mus musculus   \n",
       "2            GEO      2014         Nov 07, 2003          Homo sapiens   \n",
       "3            GEO      2016         May 06, 2013     Rattus norvegicus   \n",
       "4            GEO      2016         Mar 11, 2002          Homo sapiens   \n",
       "...          ...       ...                  ...                   ...   \n",
       "78614        GEO      2019         Mar 11, 2002          Homo sapiens   \n",
       "78615        GEO      2019                  NaN                   NaN   \n",
       "78616        GEO      2013         Mar 16, 2009          Homo sapiens   \n",
       "78618        SRA      2019  2018-03-13 01:47:12  human gut metagenome   \n",
       "78619        SRA      2019  2018-03-13 01:47:12  human gut metagenome   \n",
       "\n",
       "                      hardware sraLibrary_strategy  sraAvg_length    sraBases  \\\n",
       "0                  SAGE NlaIII                 NaN            NaN         NaN   \n",
       "1      in situ oligonucleotide                 NaN            NaN         NaN   \n",
       "2      in situ oligonucleotide                 NaN            NaN         NaN   \n",
       "3      in situ oligonucleotide                 NaN            NaN         NaN   \n",
       "4      in situ oligonucleotide                 NaN            NaN         NaN   \n",
       "...                        ...                 ...            ...         ...   \n",
       "78614  in situ oligonucleotide                 NaN            NaN         NaN   \n",
       "78615                      NaN                 NaN            NaN         NaN   \n",
       "78616  in situ oligonucleotide                 NaN            NaN         NaN   \n",
       "78618           Illumina MiSeq            AMPLICON          602.0  17743348.0   \n",
       "78619           Illumina MiSeq            AMPLICON          602.0  17743348.0   \n",
       "\n",
       "      sraAccess  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "...         ...  \n",
       "78614       NaN  \n",
       "78615       NaN  \n",
       "78616       NaN  \n",
       "78618    public  \n",
       "78619    public  \n",
       "\n",
       "[73901 rows x 13 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean up columns with factor for both SRA and GEO, rearrange columns\n",
    "allFactors['species'] = allFactors['sraSpecies'].fillna(allFactors['geoSpecies'])\n",
    "allFactors = allFactors.drop(labels = ['sraSpecies', 'geoSpecies'], axis = 1)\n",
    "\n",
    "allFactors['hardware'] = allFactors['sraHardware'].fillna(allFactors['geoHardware'])\n",
    "allFactors = allFactors.drop(labels = ['sraHardware', 'geoHardware'], axis = 1)\n",
    "\n",
    "allFactors['repository_date'] = allFactors['sraRelease'].fillna(allFactors['geoRelease'])\n",
    "allFactors = allFactors.drop(labels = ['sraRelease', 'geoRelease'], axis = 1)\n",
    "\n",
    "cols = ['journal', 'pmc_ID', 'accession', 'converted_accession', 'repository', \n",
    "        'pmc_date', 'repository_date', 'species', \n",
    "        'hardware', 'sraLibrary_strategy', 'sraAvg_length', 'sraBases', 'sraAccess']\n",
    "\n",
    "allFactors = allFactors[cols]\n",
    "allFactors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # old loop that takes forever, replaced by merge functions... holding onto just in case\n",
    "\n",
    "# ds_date = []\n",
    "# hardware = []\n",
    "# lib_type = []\n",
    "# species = []\n",
    "# tput = []\n",
    "# readLength = []\n",
    "# access = []\n",
    "\n",
    "# for acc in pmcData['converted_accession']:\n",
    "#     if(type(acc) == str):\n",
    "#         if('GSE' in acc):\n",
    "#             ds_date.append(grabRelated(geoAttributes, acc, 'Series', 'Release Date'))\n",
    "#             hardware.append(grabRelated(geoAttributes, acc, 'Series', 'Technology'))\n",
    "#             lib_type.append('MISSING')\n",
    "#             species.append(grabRelated(geoAttributes, acc, 'Series', 'Taxonomy'))\n",
    "#             tput.append('MISSING')\n",
    "#             readLength.append('MISSING')\n",
    "#             access.append('MISSING')\n",
    "#         if('SRP' in acc or 'ERP' in acc or 'DRP' in acc):\n",
    "#             ds_date.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'ReleaseDate'))\n",
    "#             hardware.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'Model'))\n",
    "#             lib_type.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'LibraryStrategy'))\n",
    "#             species.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'ScientificName'))\n",
    "#             tput.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'bases'))\n",
    "#             readLength.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'avgLength'))\n",
    "#             access.append(grabRelated(sraAttributes, acc, 'SRAStudy', 'Consent'))\n",
    "#     else:\n",
    "#         ds_date.append('NaN')\n",
    "#         hardware.append('NaN')\n",
    "#         lib_type.append('NaN')\n",
    "#         species.append('NaN')\n",
    "#         tput.append('NaN')\n",
    "#         readLength.append('NaN')\n",
    "#         access.append('NaN')\n",
    "        \n",
    "        \n",
    "# # # for reference, amt. of rows in pmcData\n",
    "# # print(len(pmcData['journal']))   \n",
    "\n",
    "# # # amt of items in each potential column\n",
    "# # for i in [ds_date, hardware, lib_type, species, tput, readLength, access]:\n",
    "# #     print(len(i))\n",
    "\n",
    "# pmcData.loc[:,'dataset_date'] = ds_date\n",
    "# pmcData.loc[:,'hardware'] = hardware\n",
    "# pmcData.loc[:,'library_type'] = lib_type\n",
    "# pmcData.loc[:,'species'] = species\n",
    "# pmcData.loc[:,'throughput'] = tput\n",
    "# pmcData.loc[:,'avg_readLength'] = readLength\n",
    "# pmcData.loc[:,'access_restrictions'] = access\n",
    "\n",
    "# pmcData\n",
    "\n",
    "# pmcData.to_csv('pmcFactorMatrix.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
