{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# create_metadata_table\n",
    "\n",
    "This notebook produces a table of PMC paper IDs matched to the SRA or GEO datasets they contain, and metadata about those datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "import statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import and clean the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import GEO reference data\n",
    "\n",
    "# import table with accession, platform, and series\n",
    "geoAPS = pd.read_csv('../geo_sample.csv')\n",
    "\n",
    "# import table with datasets IDs\n",
    "geoSeries = pd.read_csv('../geo_series_table.csv', low_memory = False)\n",
    "geoSeries.rename(columns={'Accession':'Series'}, inplace = True)\n",
    "\n",
    "# add datasets column by merging\n",
    "allData = pd.merge(geoAPS, geoSeries, how = 'outer', on = 'Series')\n",
    "geoReference = allData[['Series', 'Accession', 'Platform', 'Datasets']].drop_duplicates()\n",
    "\n",
    "# geoReference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import SRA reference data\n",
    "sraReference = pd.read_csv('../sraIDfull.csv', error_bad_lines = False, low_memory=False, quoting=3)\n",
    "sraReference = sraReference[['SRAStudy', 'Run', 'Experiment', \n",
    "                             'BioProject', 'Submission', 'Sample']].drop_duplicates()\n",
    "# sraReference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data scraped from PubMed\n",
    "# this data was scraped from XML files on the hoffman2 cluster: /u/scratch/n/nikodm/pmcOA/\n",
    "\n",
    "pmcData = pd.read_csv('../data_tables/preFilterMatrix.csv').drop_duplicates()\n",
    "# pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge GEO accessions with reference data, convert to Series\n",
    "\n",
    "# match series first\n",
    "se = pd.merge(pmcData, geoReference['Series'].drop_duplicates(), how = 'left',\n",
    "            left_on = 'accession', right_on = 'Series')\n",
    "pmcData = se.rename(columns = {'Series': 'Series_result'})\n",
    "\n",
    "# match each other style of GEO ID\n",
    "for col in ['Accession', 'Platform', 'Datasets']:\n",
    "    pmcData = pd.merge(pmcData, geoReference[['Series', col]].drop_duplicates(subset = col), how = 'left', \n",
    "            left_on = 'accession', right_on = col)\n",
    "    label = col + '_result'\n",
    "    pmcData = pmcData.rename(columns = {'Series': label})\n",
    "\n",
    "# combine all GEO series match columns into one aggregate GEO series column, clean up\n",
    "pmcData['geoSeries'] = pmcData['Series_result'].fillna(pmcData['Accession_result']).fillna(pmcData['Platform_result']).fillna(pmcData['Datasets_result'])\n",
    "pmcData = pmcData.drop(labels = ['Accession', 'Platform', 'Datasets',\n",
    "                           'Series_result', 'Accession_result',\n",
    "                           'Platform_result', 'Datasets_result'], axis = 1)\n",
    "pmcData_geoMerged = pmcData\n",
    "# pmcData_geoMerged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge SRA accessions with reference data, convert to Study\n",
    "\n",
    "# match SRA Study IDs first\n",
    "st = pd.merge(pmcData_geoMerged, sraReference['SRAStudy'].drop_duplicates(), how = 'left',\n",
    "            left_on = 'accession', right_on = 'SRAStudy')\n",
    "pmcData = st.rename(columns = {'SRAStudy': 'Study_result'})\n",
    "\n",
    "# match every other style of SRA ID\n",
    "for col in ['Run', 'Experiment', 'BioProject', 'Submission', 'Sample']:\n",
    "    pmcData = pd.merge(pmcData, sraReference[['SRAStudy', col]].drop_duplicates(subset = [col]), how = 'left',\n",
    "                      left_on = 'accession', right_on = col)\n",
    "    label = col + '_result'\n",
    "    pmcData = pmcData.rename(columns = {'SRAStudy': label})\n",
    "\n",
    "# combine all SRA Study matches into one aggregate column, clean up\n",
    "pmcData['sraStudy'] = pmcData['Study_result'].fillna(pmcData['Run_result']).fillna(pmcData['Experiment_result']).fillna(pmcData['BioProject_result']).fillna(pmcData['Submission_result']).fillna(pmcData['Sample_result'])\n",
    "pmcData = pmcData.drop(labels = ['Run', 'Experiment', 'BioProject', 'Submission', 'Sample',\n",
    "                                'Study_result', 'Run_result', 'Experiment_result', 'BioProject_result',\n",
    "                                'Submission_result', 'Sample_result'], axis = 1)\n",
    "\n",
    "pmcData_sraMerged = pmcData\n",
    "# pmcData_sraMerged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine GEO Series hits and SRA study hits into one converted accession column\n",
    "pmcData['converted_accession'] = pmcData_sraMerged['geoSeries'].fillna(pmcData_sraMerged['sraStudy'])\n",
    "pmcData = pmcData.drop(labels = ['geoSeries', 'sraStudy'], axis = 1)\n",
    "# pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean out garbage converted_accession entries, and rows that didn't map to a converted_accession\n",
    "w = []\n",
    "for a in pmcData['converted_accession']:\n",
    "    if(type(a) == str):\n",
    "        if a[0:3] != 'GSE' and a[0:3] != 'GPL' and a[1:3] != 'RP' and a not in w:\n",
    "                w.append(a)\n",
    "w.append(np.NaN)\n",
    "        \n",
    "pmcData = pmcData[~pmcData.converted_accession.isin(w)]\n",
    "# pmcData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Collect desired factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import SRA attribute data\n",
    "# CAUTION: large file! Time delay on import...\n",
    "sraAttributes = pd.read_csv('../sra_complete_runs.csv', error_bad_lines = False, low_memory=False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert SRA dates to a universal format\n",
    "pd.set_option('display.max_columns', 50)\n",
    "sraAttributes['ReleaseDate'] = sraAttributes['ReleaseDate'].str[0:10]\n",
    "# sraAttributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define functions to convert GEO dates to a universal format\n",
    "def strToMonth(m):\n",
    "    if(m == 'Jan'):\n",
    "        return '01'\n",
    "    elif(m == 'Feb'):\n",
    "        return '02'\n",
    "    elif(m == 'Mar'):\n",
    "        return '03'\n",
    "    elif(m == 'Apr'):\n",
    "        return '04'\n",
    "    elif(m == 'May'):\n",
    "        return '05'\n",
    "    elif(m == 'Jun'):\n",
    "        return '06'\n",
    "    elif(m == 'Jul'):\n",
    "        return '07'\n",
    "    elif(m == 'Aug'):\n",
    "        return '08'\n",
    "    elif(m == 'Sep'):\n",
    "        return '09'\n",
    "    elif(m == 'Oct'):\n",
    "        return '10'\n",
    "    elif(m == 'Nov'):\n",
    "        return '11'\n",
    "    elif(m == 'Dec'):\n",
    "        return '12'\n",
    "    else:\n",
    "        return(np.NaN)\n",
    "    \n",
    "def convGEODate(d):\n",
    "    if(type(d) == str):\n",
    "        mon = strToMonth(d[0:3])\n",
    "        day = d[4:6]\n",
    "        yr = d[8:12]\n",
    "        return yr + '-' + mon + '-' + day\n",
    "    else:\n",
    "        return np.NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import GEO attribute data and add Series column\n",
    "geoPlatforms = pd.read_csv('../geo_platforms_table.csv')\n",
    "geoPlatforms.rename(columns={'Accession':'Platform'}, inplace = True)\n",
    "techByPlatform = geoPlatforms[['Platform', 'Technology']]\n",
    "\n",
    "# allData contains metadata matched to GEO series, but lacks 'Technology' column\n",
    "geoAttributes = pd.merge(allData, techByPlatform, how = 'left', on = 'Platform')\n",
    "# geoAttributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert GEO dates to universal format\n",
    "dates = []\n",
    "for i in geoAttributes['Release Date']:\n",
    "    dates.append(convGEODate(i))\n",
    "geoAttributes['Release Date'] = dates\n",
    "# geoAttributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add a column tagging each accession as GEO or SRA\n",
    "\n",
    "repoList = []\n",
    "\n",
    "for i in pmcData['converted_accession']:\n",
    "    if(type(i) == str):\n",
    "        if('GSE' in i or 'GPL' in i):\n",
    "            repoList.append('GEO')\n",
    "        elif('SRP' in i or 'ERP' in i or 'DRP' in i):\n",
    "            repoList.append('SRA')\n",
    "        else:\n",
    "            repoList.append(np.NaN)\n",
    "    else:\n",
    "        repoList.append(np.NaN)\n",
    "        \n",
    "pmcData['repository'] = repoList\n",
    "# pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# add column for paper publish date\n",
    "# this data was scraped from XML files on the hoffman2 cluster: /u/scratch/n/nikodm/pmcOA/\n",
    "\n",
    "pmc_dates = pd.read_csv('../data_lists/preFilterDates.csv').drop_duplicates()\n",
    "\n",
    "pmcData = pd.merge(pmcData, pmc_dates, how = 'left', on = 'pmc_ID')\n",
    "pmcData = pmcData.rename(columns = {'date': 'pmc_date'})\n",
    "\n",
    "# pmcData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get every factor we're interested in from our tables of GEO and SRA metadata...\n",
    "\n",
    "# take a slice of the GEO and SRA attribute tables with only the info we want\n",
    "slicedGEOAtt = geoAttributes[['Series', 'Release Date', 'Technology', 'Taxonomy']]\n",
    "slicedGEOAtt.columns = ['converted_accession', 'geoRelease', 'geoHardware', 'geoSpecies']\n",
    "slicedGEOAtt = slicedGEOAtt.drop_duplicates(subset = ['converted_accession'])\n",
    "\n",
    "slicedSRAAtt = sraAttributes[['SRAStudy', 'ReleaseDate', 'Model', \n",
    "                              'LibraryStrategy', 'ScientificName', \n",
    "                              'bases', 'avgLength', 'Consent']]\n",
    "slicedSRAAtt.columns = ['converted_accession', 'sraRelease', 'sraHardware', \n",
    "                        'sraLibrary_strategy', 'sraSpecies', \n",
    "                        'sraBases', 'sraAvg_length', 'sraAccess']\n",
    "slicedSRAAtt = slicedSRAAtt.drop_duplicates(subset = ['converted_accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# special case for GEO: make an educated guess on library strategy based on hardware\n",
    "# These guesses are based on manually checking GEO series IDs that corresponded to various types of hardware\n",
    "\n",
    "gc = Counter(geoAttributes['Technology'])\n",
    "ls_guesses = pd.DataFrame.from_dict(gc, orient='index').reset_index()\n",
    "ls_guesses.columns = ['hardware', 'use_count']\n",
    "ls_guesses = ls_guesses.drop(labels = ['use_count'], axis = 1)\n",
    "\n",
    "ls = []\n",
    "\n",
    "for i in ls_guesses['hardware']:\n",
    "    if(i == 'high-throughput sequencing'):\n",
    "        ls.append('RNA-Seq')\n",
    "    elif(i == 'SAGE NlaIII' or i == 'spotted DNA/cDNA' or i == 'SAGE Sau3A' \n",
    "         or i == 'in situ oligonucleotide' or i == 'spotted oligonucleotide' \n",
    "         or i == 'antibody' or i == 'MPSS' or i == 'oligonucleotide beads' \n",
    "         or i == 'RT-PCR' or i == 'mixed spotted oligonucleotide/cDNA' \n",
    "         or i == 'spotted peptide or protein'):\n",
    "        ls.append('Expression_Array')\n",
    "    else:\n",
    "        ls.append(np.NaN)\n",
    "        \n",
    "ls_guesses.loc[:,'geoLibrary_strategy'] = ls\n",
    "# ls_guesses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge SRA attributes onto pmcData table\n",
    "mergedSRA = pd.merge(pmcData, slicedSRAAtt, how = 'left', on = 'converted_accession')\n",
    "mergedSRA = mergedSRA.drop_duplicates()\n",
    "\n",
    "# merge GEO attributes onto table of pmcData + SRA Attributes\n",
    "allFactors = pd.merge(mergedSRA, slicedGEOAtt, how = 'left', on = 'converted_accession')\n",
    "allFactors = pd.merge(allFactors, ls_guesses, how = 'left', left_on = 'geoHardware', right_on = 'hardware')\n",
    "\n",
    "allFactors = allFactors.dropna(subset = ['converted_accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "      <th>repository</th>\n",
       "      <th>pmc_date</th>\n",
       "      <th>pmc_date</th>\n",
       "      <th>repository_date</th>\n",
       "      <th>species</th>\n",
       "      <th>hardware</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>sraAvg_length</th>\n",
       "      <th>sraBases</th>\n",
       "      <th>sraAccess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013-04-18</td>\n",
       "      <td>2013-04-18</td>\n",
       "      <td>2003-07-16</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013-05-14</td>\n",
       "      <td>2013-05-14</td>\n",
       "      <td>2013-03-28</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2014-11-02</td>\n",
       "      <td>2014-11-02</td>\n",
       "      <td>2006-07-10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016-01-28</td>\n",
       "      <td>2016-01-28</td>\n",
       "      <td>2016-03-17</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016-01-28</td>\n",
       "      <td>2016-01-28</td>\n",
       "      <td>2004-07-16</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>223959</td>\n",
       "      <td>Saudi_J_Biol_Sci</td>\n",
       "      <td>PMC6088103</td>\n",
       "      <td>GSE28146</td>\n",
       "      <td>GSE28146</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2018-05-18</td>\n",
       "      <td>2018-05-18</td>\n",
       "      <td>2011-08-01</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>223960</td>\n",
       "      <td>Saudi_J_Biol_Sci</td>\n",
       "      <td>PMC6933160</td>\n",
       "      <td>ERS2923991</td>\n",
       "      <td>ERP109196</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019-05-07</td>\n",
       "      <td>2019-05-07</td>\n",
       "      <td>2019-01-24</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>454 GS FLX</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>250.0</td>\n",
       "      <td>6461726.0</td>\n",
       "      <td>public</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>223961</td>\n",
       "      <td>Saudi_J_Biol_Sci</td>\n",
       "      <td>PMC6933160</td>\n",
       "      <td>ERS2923992</td>\n",
       "      <td>ERP109196</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019-05-07</td>\n",
       "      <td>2019-05-07</td>\n",
       "      <td>2019-01-24</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>454 GS FLX</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>250.0</td>\n",
       "      <td>6461726.0</td>\n",
       "      <td>public</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>223962</td>\n",
       "      <td>Saudi_J_Biol_Sci</td>\n",
       "      <td>PMC6933160</td>\n",
       "      <td>ERS2923995</td>\n",
       "      <td>ERP109196</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019-05-07</td>\n",
       "      <td>2019-05-07</td>\n",
       "      <td>2019-01-24</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>454 GS FLX</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>250.0</td>\n",
       "      <td>6461726.0</td>\n",
       "      <td>public</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>223963</td>\n",
       "      <td>Saudi_Pharm_J</td>\n",
       "      <td>PMC6218847</td>\n",
       "      <td>GPL8</td>\n",
       "      <td>GSE2</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2018-05-24</td>\n",
       "      <td>2018-05-24</td>\n",
       "      <td>2001-04-26</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>spotted DNA/cDNA</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223964 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    journal      pmc_ID   accession converted_accession  \\\n",
       "0       Alzheimers_Res_Ther  PMC3707052        GSM1              GSE506   \n",
       "1       Alzheimers_Res_Ther  PMC3706879    GSE45534            GSE45534   \n",
       "2       Alzheimers_Res_Ther  PMC4255636     GSE5281             GSE5281   \n",
       "3       Alzheimers_Res_Ther  PMC4731966    GSE67036            GSE67036   \n",
       "4       Alzheimers_Res_Ther  PMC4731966     GSE1297             GSE1297   \n",
       "...                     ...         ...         ...                 ...   \n",
       "223959     Saudi_J_Biol_Sci  PMC6088103    GSE28146            GSE28146   \n",
       "223960     Saudi_J_Biol_Sci  PMC6933160  ERS2923991           ERP109196   \n",
       "223961     Saudi_J_Biol_Sci  PMC6933160  ERS2923992           ERP109196   \n",
       "223962     Saudi_J_Biol_Sci  PMC6933160  ERS2923995           ERP109196   \n",
       "223963        Saudi_Pharm_J  PMC6218847        GPL8                GSE2   \n",
       "\n",
       "       repository    pmc_date    pmc_date repository_date            species  \\\n",
       "0             GEO  2013-04-18  2013-04-18      2003-07-16       Homo sapiens   \n",
       "1             GEO  2013-05-14  2013-05-14      2013-03-28       Mus musculus   \n",
       "2             GEO  2014-11-02  2014-11-02      2006-07-10       Homo sapiens   \n",
       "3             GEO  2016-01-28  2016-01-28      2016-03-17  Rattus norvegicus   \n",
       "4             GEO  2016-01-28  2016-01-28      2004-07-16       Homo sapiens   \n",
       "...           ...         ...         ...             ...                ...   \n",
       "223959        GEO  2018-05-18  2018-05-18      2011-08-01       Homo sapiens   \n",
       "223960        SRA  2019-05-07  2019-05-07      2019-01-24         metagenome   \n",
       "223961        SRA  2019-05-07  2019-05-07      2019-01-24         metagenome   \n",
       "223962        SRA  2019-05-07  2019-05-07      2019-01-24         metagenome   \n",
       "223963        GEO  2018-05-24  2018-05-24      2001-04-26       Mus musculus   \n",
       "\n",
       "                       hardware  library_strategy  sraAvg_length   sraBases  \\\n",
       "0                   SAGE NlaIII  Expression_Array            NaN        NaN   \n",
       "1       in situ oligonucleotide  Expression_Array            NaN        NaN   \n",
       "2       in situ oligonucleotide  Expression_Array            NaN        NaN   \n",
       "3       in situ oligonucleotide  Expression_Array            NaN        NaN   \n",
       "4       in situ oligonucleotide  Expression_Array            NaN        NaN   \n",
       "...                         ...               ...            ...        ...   \n",
       "223959  in situ oligonucleotide  Expression_Array            NaN        NaN   \n",
       "223960               454 GS FLX          AMPLICON          250.0  6461726.0   \n",
       "223961               454 GS FLX          AMPLICON          250.0  6461726.0   \n",
       "223962               454 GS FLX          AMPLICON          250.0  6461726.0   \n",
       "223963         spotted DNA/cDNA  Expression_Array            NaN        NaN   \n",
       "\n",
       "       sraAccess  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "...          ...  \n",
       "223959       NaN  \n",
       "223960    public  \n",
       "223961    public  \n",
       "223962    public  \n",
       "223963       NaN  \n",
       "\n",
       "[223964 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clean up columns with factor for both SRA and GEO, rearrange columns\n",
    "allFactors['species'] = allFactors['sraSpecies'].fillna(allFactors['geoSpecies'])\n",
    "allFactors = allFactors.drop(labels = ['sraSpecies', 'geoSpecies'], axis = 1)\n",
    "\n",
    "allFactors['hardware'] = allFactors['sraHardware'].fillna(allFactors['geoHardware'])\n",
    "allFactors = allFactors.drop(labels = ['sraHardware', 'geoHardware'], axis = 1)\n",
    "\n",
    "allFactors['library_strategy'] = allFactors['sraLibrary_strategy'].fillna(allFactors['geoLibrary_strategy'])\n",
    "allFactors = allFactors.drop(labels = ['sraLibrary_strategy', 'geoLibrary_strategy'], axis = 1)\n",
    "\n",
    "allFactors['repository_date'] = allFactors['sraRelease'].fillna(allFactors['geoRelease'])\n",
    "allFactors = allFactors.drop(labels = ['sraRelease', 'geoRelease'], axis = 1)\n",
    "\n",
    "cols = ['journal', 'pmc_ID', 'accession', 'converted_accession', 'repository', \n",
    "        'pmc_date', 'repository_date', 'species', \n",
    "        'hardware', 'library_strategy', 'sraAvg_length', 'sraBases', 'sraAccess']\n",
    "\n",
    "allFactors = allFactors[cols]\n",
    "allFactors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to .csv\n",
    "allFactors.to_csv('../data_tables/metadataMatrix_raw.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
