{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>journal</th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>accession</th>\n",
       "      <th>converted_accession</th>\n",
       "      <th>repository</th>\n",
       "      <th>pmc_date</th>\n",
       "      <th>repository_date</th>\n",
       "      <th>species</th>\n",
       "      <th>hardware</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>sraAvg_length</th>\n",
       "      <th>sraBases</th>\n",
       "      <th>sraAccess</th>\n",
       "      <th>total_use_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3707052</td>\n",
       "      <td>GSM1</td>\n",
       "      <td>GSE506</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "      <td>Sep 28, 2000</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SAGE NlaIII</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC3706879</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GSE45534</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "      <td>May 25, 2004</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4255636</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GSE5281</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2014</td>\n",
       "      <td>Nov 07, 2003</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GSE67036</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016</td>\n",
       "      <td>May 06, 2013</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Alzheimers_Res_Ther</td>\n",
       "      <td>PMC4731966</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GSE1297</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2016</td>\n",
       "      <td>Mar 11, 2002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73928</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589307</td>\n",
       "      <td>GSE14795</td>\n",
       "      <td>GSE14795</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2019</td>\n",
       "      <td>Mar 11, 2002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73929</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC6589301</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GSE62254</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73930</td>\n",
       "      <td>Dis_Markers</td>\n",
       "      <td>PMC3834650</td>\n",
       "      <td>GDS1059</td>\n",
       "      <td>GSE2191</td>\n",
       "      <td>GEO</td>\n",
       "      <td>2013</td>\n",
       "      <td>Mar 16, 2009</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>Expression_Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73931</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>PRJNA437812</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019</td>\n",
       "      <td>2018-03-13 01:47:12</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>602.0</td>\n",
       "      <td>17743348.0</td>\n",
       "      <td>public</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73932</td>\n",
       "      <td>Case_Rep_Hematol</td>\n",
       "      <td>PMC6462343</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRP135146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>2019</td>\n",
       "      <td>2018-03-13 01:47:12</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>602.0</td>\n",
       "      <td>17743348.0</td>\n",
       "      <td>public</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73933 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   journal      pmc_ID    accession converted_accession  \\\n",
       "0      Alzheimers_Res_Ther  PMC3707052         GSM1              GSE506   \n",
       "1      Alzheimers_Res_Ther  PMC3706879     GSE45534            GSE45534   \n",
       "2      Alzheimers_Res_Ther  PMC4255636      GSE5281             GSE5281   \n",
       "3      Alzheimers_Res_Ther  PMC4731966     GSE67036            GSE67036   \n",
       "4      Alzheimers_Res_Ther  PMC4731966      GSE1297             GSE1297   \n",
       "...                    ...         ...          ...                 ...   \n",
       "73928          Dis_Markers  PMC6589307     GSE14795            GSE14795   \n",
       "73929          Dis_Markers  PMC6589301     GSE62254            GSE62254   \n",
       "73930          Dis_Markers  PMC3834650      GDS1059             GSE2191   \n",
       "73931     Case_Rep_Hematol  PMC6462343  PRJNA437812           SRP135146   \n",
       "73932     Case_Rep_Hematol  PMC6462343    SRP135146           SRP135146   \n",
       "\n",
       "      repository  pmc_date      repository_date               species  \\\n",
       "0            GEO      2013         Sep 28, 2000          Homo sapiens   \n",
       "1            GEO      2013         May 25, 2004          Mus musculus   \n",
       "2            GEO      2014         Nov 07, 2003          Homo sapiens   \n",
       "3            GEO      2016         May 06, 2013     Rattus norvegicus   \n",
       "4            GEO      2016         Mar 11, 2002          Homo sapiens   \n",
       "...          ...       ...                  ...                   ...   \n",
       "73928        GEO      2019         Mar 11, 2002          Homo sapiens   \n",
       "73929        GEO      2019                  NaN                   NaN   \n",
       "73930        GEO      2013         Mar 16, 2009          Homo sapiens   \n",
       "73931        SRA      2019  2018-03-13 01:47:12  human gut metagenome   \n",
       "73932        SRA      2019  2018-03-13 01:47:12  human gut metagenome   \n",
       "\n",
       "                      hardware  library_strategy  sraAvg_length    sraBases  \\\n",
       "0                  SAGE NlaIII  Expression_Array            NaN         NaN   \n",
       "1      in situ oligonucleotide  Expression_Array            NaN         NaN   \n",
       "2      in situ oligonucleotide  Expression_Array            NaN         NaN   \n",
       "3      in situ oligonucleotide  Expression_Array            NaN         NaN   \n",
       "4      in situ oligonucleotide  Expression_Array            NaN         NaN   \n",
       "...                        ...               ...            ...         ...   \n",
       "73928  in situ oligonucleotide  Expression_Array            NaN         NaN   \n",
       "73929                      NaN               NaN            NaN         NaN   \n",
       "73930  in situ oligonucleotide  Expression_Array            NaN         NaN   \n",
       "73931           Illumina MiSeq          AMPLICON          602.0  17743348.0   \n",
       "73932           Illumina MiSeq          AMPLICON          602.0  17743348.0   \n",
       "\n",
       "      sraAccess  total_use_count  \n",
       "0           NaN               64  \n",
       "1           NaN                1  \n",
       "2           NaN               43  \n",
       "3           NaN                1  \n",
       "4           NaN               32  \n",
       "...         ...              ...  \n",
       "73928       NaN                1  \n",
       "73929       NaN               22  \n",
       "73930       NaN                5  \n",
       "73931    public                1  \n",
       "73932    public                1  \n",
       "\n",
       "[73933 rows x 14 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import metadata matrix\n",
    "mm = pd.read_csv('../data_tables/metadataMatrix.csv')\n",
    "mm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function defining each paper as a GEO or SRA paper...\n",
    "# ...based on which is mentioned more, or 'Both' if score is equal\n",
    "def getRepo(paper_IDs):\n",
    "    try:\n",
    "        r = statistics.mode(paper_IDs)\n",
    "    except:\n",
    "        r = 'Both'\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmc_ID</th>\n",
       "      <th>repository</th>\n",
       "      <th>avg_use</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>PMC1079818</td>\n",
       "      <td>Both</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>PMC1082884</td>\n",
       "      <td>Both</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>PMC1088941</td>\n",
       "      <td>Both</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PMC1088946</td>\n",
       "      <td>Both</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>PMC1090574</td>\n",
       "      <td>Both</td>\n",
       "      <td>1.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26065</td>\n",
       "      <td>PMC6639121</td>\n",
       "      <td>Both</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26066</td>\n",
       "      <td>PMC6639415</td>\n",
       "      <td>Both</td>\n",
       "      <td>1.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26067</td>\n",
       "      <td>PMC6640630</td>\n",
       "      <td>Both</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26068</td>\n",
       "      <td>PMC6646343</td>\n",
       "      <td>Both</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26069</td>\n",
       "      <td>PMC6646359</td>\n",
       "      <td>Both</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26070 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pmc_ID repository   avg_use\n",
       "0      PMC1079818       Both  1.000000\n",
       "1      PMC1082884       Both  2.000000\n",
       "2      PMC1088941       Both  2.000000\n",
       "3      PMC1088946       Both  2.000000\n",
       "4      PMC1090574       Both  1.400000\n",
       "...           ...        ...       ...\n",
       "26065  PMC6639121       Both  2.000000\n",
       "26066  PMC6639415       Both  1.666667\n",
       "26067  PMC6640630       Both  1.000000\n",
       "26068  PMC6646343       Both  1.000000\n",
       "26069  PMC6646359       Both  1.000000\n",
       "\n",
       "[26070 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at each PAPER\n",
    "# Across GEO and SRA, what percentage of papers reuse data? (stacked bar chart)\n",
    "\n",
    "# Take a slice of allFactors with the data we want\n",
    "byPaper = mm[['pmc_ID', 'repository', 'total_use_count']]\n",
    "byPaper = byPaper.dropna(subset = ['repository'])\n",
    "byPaper = byPaper.groupby(['pmc_ID'], as_index = False).agg({\n",
    "    'repository': lambda x: getRepo(x), 'total_use_count': 'mean'})\n",
    "byPaper.columns = ['pmc_ID', 'repository', 'avg_use']\n",
    "\n",
    "# divide into GEO, SRA, Both\n",
    "byPaperGEO = byPaper.loc[byPaper['repository'] == 'GEO']\n",
    "byPaperSRA = byPaper.loc[byPaper['repository'] == 'SRA']\n",
    "byPaperBOTH = byPaper.loc[byPaper['repository'] == 'Both']\n",
    "\n",
    "byPaper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tc = Counter(byPaper['avg_use'])\n",
    "totalCounts = pd.DataFrame.from_dict(tc, orient = 'index').reset_index()\n",
    "\n",
    "gc = Counter(byPaperGEO['avg_use'])\n",
    "geoCounts = pd.DataFrame.from_dict(gc, orient = 'index').reset_index()\n",
    "\n",
    "sc = Counter(byPaperSRA['avg_use'])\n",
    "sraCounts = pd.DataFrame.from_dict(sc, orient = 'index').reset_index()\n",
    "\n",
    "bc = Counter(byPaperBOTH['avg_use'])\n",
    "bothCounts = pd.DataFrame.from_dict(bc, orient = 'index').reset_index()\n",
    "\n",
    "for df in [totalCounts, geoCounts, sraCounts, bothCounts]:\n",
    "    df.columns = ['avg_use', 'num_papers']\n",
    "    df = df.sort_values(by = 'times_used').reset_index().drop(labels = ['index'], axis = 1)\n",
    "    # no_reuse is number of papers only mentioning datasets used once\n",
    "    no_reuse = df['num_papers'][0]\n",
    "    # tot_papers includes all papers, including those that mentioned at least one reused dataset\n",
    "    tot_papers = df['num_papers'].sum()\n",
    "    pct_reuse_papers = 1 - no_reuse/tot_papers\n",
    "    print(pct_reuse_papers)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
